main.md

```diff
@@ -401,6 +435,91 @@ known type but containing unknown elements or elements of the wrong type, the AS
+is asking for "more" or "less" than a previous, existing request. For example, upon refreshing a token, the client can
+ask for a new access token with "fewer permissions" than had been previously authorized by the resource owner.
+Since the nature of an authorization details request is based solely on the API or APIs that it is describing, there is not
+a simple means of comparing any two arbitrary authorization details requests. 
+Authorization servers should not rely on simple object comparison in most cases, as the intersection of some elements
+within a request could have side effects in the access rights granted, depending on how the API
+has been designed and deployed. This is a similar effect to the scope values used with some APIs.
+
+However, when comparing a new request to an existing request, authorization servers can use the same 
+processing techniques as used in granting the request in the first place to determine if a resource
+owner needs to authorize the request. The details of this comparison are dependent on the definition
+of the `type` of authorization request and outside the scope of this specification, but common patterns
+can be applied.
+
+This shall be illustrated using our running example. The example authorization request in (#authz_request), if approved by the user, resulted in the issuance of an authorization code associated with the privileges to 
+
+* list accounts 
+* access the balance of one or more accounts,
+* access the transactions of one or more accounts, and 
+* to initiate a payment. 
+
+The client could now request the AS to issue an access token assigned with the privilege to just access a list of accounts as follows:
+
+```JSON
+[
+   {
+      "type":"account_information",
+      "actions":[
+         "list_accounts"
+      ],
+      "locations":[
+         "https://example.com/accounts"
+      ]
+   }
+]
+```
+
+The example API is designed such that each field used by the `account_information` type contains additive rights, 
+with each value within the `actions` and `locations` arrays specifying a different element of access. To make a comparison in this
+instance, the AS would perform the following steps:
+
+* compare that the authorization code issued in the previous step contains an authorization details object of type `account_information`
+* compare whether the approved list of actions contains `list_account`, and 
+* whether the `locations` value includes only previously-approved locations.
+
+If all checks succeed, the AS would issue the requested access token with the reduced set of access. 
+
+Note that this comparison is relevant A different API type definition could have different processing rules. For example, the
+value of an `action` could subsume the rights associated with another `action` value. For example, if a client initially asks for
+a token with `write` access, which implies both read and write access to this API:
```

Comment:
It is easy to drop existing entries to narrow down actions. For example, it is easy to narrow down actions from `["read","write"]` to `["read"]`. However, it is difficult for AS implementations to understand that `write` subsumes `read`. To achieve the subsumption, the AS must have a mechanism to manage subsumption relationships among actions and enable AS administrators to configure the relationships. Such mechanism is too heavy compared to the benefits gained.

Reply:
Such mechanisms are up to the AS to implement, though, and the protocol is silent on them. This is true for `scope`, `resource`, and should remain true for RAR.

Reply:
I've been trying to express similar sentiment as Taka on the difficulty vs. benefit of arbitrary comparison of authz deets but the train seems to be rolling out regardless.

Reply:
I agree that it's difficult but I don't think we get to pick if people try to do it a hard way, unless we want to get deep into defining semantics across all APIs. 

Also, keep in mind that our examples here are only using the "internal" fields, an API type could have its own completely separate field with completely unique comparison semantics. Maybe it's a bitfield? Maybe it's an s-expression with recursive calculations on risk in context? Plus these types could have additional semantics at the object level: maybe asking for action `foo` at location `A` is "more" than asking for action `bar` at location `B`, according to some weird graph API.

Reply:
I gained new knowledge. Different authors have different preferences. Some prefer to keep specifications abstract for flexibility and extensibility, while others prefer to make specifications deterministic for interoperability and unambiguity.

The former don't care that some AS implementations may accept `["read"]` but others may not, while the latter expect that all AS implementations generate the same output when input is identical.

A branch point would be whether or not we want the conformance suite to test the subsumption rules. The more test cases for RAR we want to have, the more deterministic the specification needs to be.

Reply:
It's not so much a difference in author preferences here -- it's the fact that this specification is ultimately about enabling API designers to describe the security options of their APIs. If this spec were to take a hardline stance on API design, the result would be API designers choosing not to use this to describe their APIs, and possibly inventing some competing mechanism. The `scope` parameter has benefited from having a very simple definition (https://tools.ietf.org/html/rfc6749#section-3.3) of being additive:

> If the value contains multiple space-delimited strings, their order does not matter, and each string adds an additional access range to the requested scope.

This very loose definition allows the AS to interpret the scope however it needs to for the API being protected. Plus, I'd argue that nobody takes that line seriously in practice, anyway, since there are examples of scope definitions being used to :limit: the access and not expand it, and the world works fine with that. 

Ultimately, anything being used to describe an API needs to be tested in the context of the API it's protecting. The OpenID Connect test suite doesn't just test `scope`, it specifically tests the OpenID Connect scope values of `openid address email phone profile offline_access` to make sure those do what they're supposed to in the context of the OpenID Connect API. The same thing is going to be true for any RAR `type` definition.

Reply:
Interoperability and flexibility are both important aspects. OAuth is so successful because of its tremendous flexibility while providing almost no interoperability (one of the reasons we need FAPI and I hope GNAP will be much better wrt interoperability). 

Interoperability for RAR is achieved on two levels: the RAR spec itself can achieve interoperability on the level of the mechanics to request and share any kind of fine grained authorization. Interoperability on the level of a certain authorization details, however, type is tied to the respective API and out of scope of RAR. Otherwise I‘m afraid to limit flexibility to much.

Reply:
@TakahikoKawasaki @b---c can we resolve this conversation?